stl/inc/mdspan

```diff
@@ -24,151 +24,126 @@ _STL_DISABLE_CLANG_WARNINGS
-    static_assert(((_Extents == dynamic_extent || _STD in_range<_IndexType>(_Extents)) && ...),
-        "Each element of Extents must be either equal to dynamic_extent, or must be representable as a value of type "
-        "IndexType (N4928 [mdspan.extents.overview]/1.2).");
-
-    _NODISCARD static constexpr rank_type rank() noexcept {
-        return sizeof...(_Extents);
-    }
-
     _NODISCARD static constexpr rank_type rank_dynamic() noexcept {
-        return ((_Extents == dynamic_extent) + ... + 0);
+        return _Rank_dynamic;
     }
 
     _NODISCARD static constexpr size_t static_extent(const rank_type _Idx) noexcept {
-        return _Mybase::_Static_extents[_Idx];
+        _STL_VERIFY(_Idx < rank(), "Index must be less than rank() (N4928 [mdspan.extents.obs]/1)");
+        return _Static_extents[_Idx];
     }
 
     _NODISCARD constexpr index_type extent(const rank_type _Idx) const noexcept {
+        _STL_VERIFY(_Idx < rank(), "Index must be less than rank() (N4928 [mdspan.extents.obs]/3)");
         if constexpr (rank_dynamic() == 0) {
-            return static_cast<index_type>(_Mybase::_Static_extents[_Idx]);
+            return static_cast<index_type>(static_extent(_Idx));
         } else if constexpr (rank_dynamic() == rank()) {
-            return _Mybase::_Dynamic_extents[_Idx];
+            return _Dynamic_extents[_Idx];
         } else {
-            const auto _Static_extent = _Mybase::_Static_extents[_Idx];
-            if (_Static_extent == dynamic_extent) {
-                return _Mybase::_Dynamic_extents[_Mybase::_Dynamic_indices[_Idx]];
+            if (static_extent(_Idx) == dynamic_extent) {
+                return _Dynamic_extents[_Dynamic_index(_Idx)];
             } else {
-                return static_cast<index_type>(_Static_extent);
+                return static_cast<index_type>(static_extent(_Idx));
             }
         }
     }
 
     constexpr extents() noexcept = default;
 
     template <class _OtherIndexType, size_t... _OtherExtents>
-        requires (sizeof...(_OtherExtents) == sizeof...(_Extents))
+        requires (sizeof...(_OtherExtents) == rank())
               && ((_OtherExtents == dynamic_extent || _Extents == dynamic_extent || _OtherExtents == _Extents) && ...)
     constexpr explicit((((_Extents != dynamic_extent) && (_OtherExtents == dynamic_extent)) || ...)
                        || (numeric_limits<index_type>::max)() < (numeric_limits<_OtherIndexType>::max)())
         extents(const extents<_OtherIndexType, _OtherExtents...>& _Other) noexcept {
-        auto _Dynamic_it = _Mybase::_Begin_dynamic_extents();
-        for (rank_type _Dim = 0; _Dim < sizeof...(_Extents); ++_Dim) {
-            if (_Mybase::_Static_extents[_Dim] == dynamic_extent) {
-                *_Dynamic_it++ = _Other.extent(_Dim);
+        auto _It = _Dynamic_extents.begin();
+        for (rank_type _Idx = 0; _Idx < rank(); ++_Idx) {

```

Comment:
No change requested but you can followup if you want: Because `rank()` is an ordinary `constexpr` function (not `consteval`), I note that usage of it outside constant expressions will result in real function calls being emitted in non-optimized debug mode. This is more of an implementer's choice than anything critical, so I'm inclined to defer to your judgement about what's clearest. However, if I were writing this, I would probably introduce a `static constexpr _Rank` data member, so I could bake it into all of these functions without real calls, while keeping the clarity of saying "rank".